lib-core/src/core-models/BaseEntity.ts

```diff
@@ -0,0 +1,105 @@
+import { Toolkit } from '../Toolkit';
+import * as _ from 'lodash';
+import { DiagramEngine } from '../DiagramEngine';
+import { BaseEvent, BaseListener, BaseObserver } from '../core/BaseObserver';
+
+export interface BaseEntityEvent<T extends BaseEntity = BaseEntity> extends BaseEvent {
+	entity: T;
+}
+
+export interface BaseEntityListener<T extends BaseEntity = BaseEntity> extends BaseListener {
+	lockChanged?(event: BaseEntityEvent<T> & { locked: boolean }): void;
+}
+
+export type BaseEntityType = 'node' | 'link' | 'port' | 'point';
+
+export interface BaseEntityOptions {
+	id?: string;
+	locked?: boolean;
+}
+
+export type BaseEntityGenerics = {
+	LISTENER: BaseEntityListener;
+	OPTIONS: BaseEntityOptions;
+};
+
+export class BaseEntity<T extends BaseEntityGenerics = BaseEntityGenerics> extends BaseObserver<T['LISTENER']> {
+	protected options: T['OPTIONS'];
+
+	constructor(options: T['OPTIONS'] = {}) {
+		super();
+		this.options = {
+			id: Toolkit.UID(),
+			...options
+		};
+	}
+
+	getOptions() {
+		return this.options;
+	}
+
+	getID() {
+		return this.options.id;
+	}
+
+	doClone(lookupTable: { [s: string]: any } = {}, clone: any) {
+		/*noop*/
+	}
+
+	clone(lookupTable: { [s: string]: any } = {}) {
+		// try and use an existing clone first
+		if (lookupTable[this.options.id]) {
+			return lookupTable[this.options.id];
+		}
+		let clone = _.clone(this);
+		clone.options = {
+			...this.options,
+			id: Toolkit.UID()
+		};
+		clone.clearListeners();
+		lookupTable[this.options.id] = clone;
+
+		this.doClone(lookupTable, clone);
+		return clone;
+	}
+
+	clearListeners() {
+		this.listeners = {};
+	}
+
+	deSerialize(data: { [s: string]: any }, engine: DiagramEngine) {
+		this.options.id = data.id;
+		this.options.locked = data.locked;
+	}
+
+	serialize(): object {
+		return {
+			id: this.options.id,
+			locked: this.options.locked
+		};
+	}
```

Comment:
There is a better way to have type-safety in `deSerialize` and `serialize`.
[TypeScript playground link](https://www.typescriptlang.org/play/index.html#code/PTAECEEEGUFFQMIBkbQLACgDGAbAhgM4ESECmoA3pqDaAaQE4CWeOTAXqQBQCUl1tQQ1IAXAK4MAdvwyC5tSXgC2pAFygARBoA0A+YIAe6gAy7Z+wQE8TeuQF8A3LbuZbAE1L1mrDtzd4RPHUAJVEJSQAVSwAHUgAeEQALJgIAbQ0vFjZODQBdAD4edQA3AHsmNxlnTBcMTA9cPGFQLFLJAhFQACMySHVwMic67DaOukYs3zcAEQC8AfpQAF5u3oA6TJ9OXiGe+kg1j03s7mOp2cCF0h4hzBBQadhggEkANVhpxBRodGx8IlA0CUSVApAMIlIkjcxCuMkEZ22fCo5nkwnEUjhFloaxxBDEsQYGwmW24PDMWNoiVITAA5okRCZyRTQAB3CpJGwo+xDQS1dyeYknLj+QIhMJSKKxBLJNIZQW+PKFErlSrI6oYPkYBr4ZqtdqdAjAxJ9QFG24jfXjbwnGZzIEglaGpIHBGkoZO42HAXW3xcV22wL2xI3TBAA)

In this case:

```
class BaseEntity {
    serialize() {
        return {
            id: this.options.id,
            locked: this.options.locked
        };
    }

    deSerialize(data: ReturnType<this["serialize"]>): void {
        this.options.id = data.id;
        this.options.locked = data.locked;
    }
}
```